### PR TITLE
support an explicit --select command line option

### PR DESCRIPTION
**Pull request by plonibarploni**

### CODE REVIEW

**Comment by cwendling** [comment]
Nautilus' `-s` is supposed to select the *basename* in the *parent* directory.  e.g.:
```console
$ nautilus /tmp
opens Nautilus displaying the contents of "/tmp"
caja works the same
$ nautilus -s /tmp
opens Nautilus displaying the contents of "/", with "tmp" selected
$ nautilus -s ~/somefile
opens Nautilus in ~ with "somefile" selected -- it works with a file as well then
```

A related feature that seems missing from Caja is that passing a *file* (and not a directory) doesn't work (gives an error with "not a directory") while Nautilus displays the containing directory scrolled to show the file:
```console
$ nautilus ~/somefile
opens Nautilus in ~ with "somefile" visible
$ caja ~/somefile
shows an error dialog
```

So without looking at the code at all, I think the feature itself makes sense and is not a duplicate.

**Comment by lukefromdc** [comment]
I can confirm that pasing a filename works with this,showing the
directory containing the file

On 4/7/2020 at 3:55 AM, "Colomban Wendling"  wrote:
	Nautilus' -s is supposed to select the basename in the parent
directory.  e.g.: 

	$ nautilus /tmp opens Nautilus displaying the contents of "/tmp" caja
works the same $ nautilus -s /tmp opens Nautilus displaying the
contents of "/", with "tmp" selected $ nautilus -s ~/somefile opens
Nautilus in ~ with "somefile" selected -- it works with a file as well
then 

	A related feature that seems missing from Caja is that passing a file
(and not a directory) doesn't work (gives an error with "not a
directory") while Nautilus displays the containing directory scrolled
to show the file: 

	$ nautilus ~/somefile opens Nautilus in ~ with "somefile" visible $
caja ~/somefile shows an error dialog 

	So without looking at the code at all, I think the feature itself
makes sense and is not a duplicate.

File: src/caja-application.c [hunk 1]
```diff
@@ -1806,6 +1807,12 @@ do_cmdline_sanity_checks (CajaApplication *self,
         goto out;
     }
 
+    if (select_uris && remaining == NULL) {
+		g_printerr ("%s\n",
+			    _("--select must be used with at least an URI."));
+		goto out;
+	}
+
```
Comment (rbuj) [inline src/caja-application.c:1815]: Use spaces instead of tabs for the indent.

**Comment by sc0w** [comment]
@plonibarploni 

which is the nautilus commit(s)?

File: src/caja-application.c [hunk 2]
```diff
@@ -1830,6 +1837,55 @@ do_perform_self_checks (gint *exit_status)
...
+	g_dbus_connection_call (connection,
+				CAJA_FDO_DBUS_NAME,
+				CAJA_FDO_DBUS_PATH,
+				CAJA_FDO_DBUS_IFACE,
+				"ShowItems",
+				g_variant_new ("(ass)", &builder, ""), NULL,
+				G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL,
+				select_items_ready_cb, self);
+
+	g_variant_builder_clear (&builder);
```
Comment (rbuj) [inline src/caja-application.c:1886]: Use spaces instead of tabs for the indent.

**Comment by plonibarploni** [comment]
> which is the nautilus commit(s)?

https://github.com/GNOME/nautilus/commit/dc7e851ae4a223522c3158dd4134a3b313283596

**Comment by raveit65** [comment]
@plonibarploni 
Can you please change commit message?
`ported from nautilus` isn't the main function of the commit.
Maybe use original commit message?
And a reference link in commit body.

**Comment by raveit65** [comment]
@plonibarploni 
I like the idea to replace tabs with spaces :)
But is there a chance to move this format fixes in a second commit without breaking your PR?
Because 50% of this commit are format changes.

File: src/caja-application.c [hunk 3]
```diff
@@ -446,10 +446,10 @@ caja_application_init (CajaApplication *application)
 
     g_action_map_add_action (G_ACTION_MAP (application), G_ACTION (action));
 
-	g_signal_connect_swapped (action, "activate",
-				  G_CALLBACK (caja_application_quit), application);
+    g_signal_connect_swapped (action, "activate",
+                  G_CALLBACK (caja_application_quit), application);
```
Comment (raveit65) [inline src/caja-application.c:450]: Indent issue. `G_CALLBACK` should start after the bracket under `action` of the line before.

File: src/caja-application.c [hunk 4]
```diff
@@ -721,7 +721,7 @@ caja_application_create_desktop_windows (CajaApplication *application)
         caja_application_desktop_windows =
             g_list_prepend (caja_application_desktop_windows, window);
             gtk_application_add_window (GTK_APPLICATION (application),
-							    GTK_WINDOW (window));
+                                GTK_WINDOW (window));
```
Comment (raveit65) [inline src/caja-application.c:724]: Indent, `GTK_WINDOW` should start under `GTK_APPLICATION` of the line before.

File: src/caja-application.c [hunk 5]
```diff
@@ -942,7 +942,7 @@ create_window (CajaApplication *application,
                            G_CONNECT_AFTER);
 
     gtk_application_add_window (GTK_APPLICATION (application),
-				    GTK_WINDOW (window));
+                    GTK_WINDOW (window));
```
Comment (raveit65) [inline src/caja-application.c:945]: Indent issue....

File: src/caja-application.c [hunk 6]
```diff
@@ -1830,6 +1837,55 @@ do_perform_self_checks (gint *exit_status)
     *exit_status = EXIT_SUCCESS;
 }
 
+static void
+select_items_ready_cb (GObject *source,
+               GAsyncResult *res,
+               gpointer user_data)
```
Comment (raveit65) [inline src/caja-application.c:1843]: Thanks, but in new or changed code the indents are wrong.
Sigh, i didn't mentioned that explicitly before.

File: src/caja-application.c [hunk 7]
```diff
@@ -1830,6 +1837,55 @@ do_perform_self_checks (gint *exit_status)
...
+
+        /* open default location instead */
+        g_application_open (G_APPLICATION (self), NULL, 0, "");
+    }
+}
+
+static void
+caja_application_select (CajaApplication *self,
+                 GFile **files,
+                 gint len)
```
Comment (raveit65) [inline src/caja-application.c:1863]: Same here, indents.

File: src/caja-application.c [hunk 8]
```diff
@@ -1830,6 +1837,55 @@ do_perform_self_checks (gint *exit_status)
...
+    }
+
+    g_dbus_connection_call (connection,
+                CAJA_FDO_DBUS_NAME,
+                CAJA_FDO_DBUS_PATH,
+                CAJA_FDO_DBUS_IFACE,
+                "ShowItems",
+                g_variant_new ("(ass)", &builder, ""), NULL,
+                G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL,
+                select_items_ready_cb, self);
```
Comment (raveit65) [inline src/caja-application.c:1884]: ....indents.

File: src/caja-application.c [hunk 9]
```diff
@@ -1929,7 +1988,7 @@ caja_application_local_command_line (GApplication *application,
     }
 
     if (!do_cmdline_sanity_checks (self, perform_self_check,
-                       version, kill_shell, remaining)) {
+                       version, kill_shell, select_uris, remaining)) {
```
Comment (raveit65) [inline src/caja-application.c:1991]: ....indents.

**Pull request merged**